daemon/logger/rollover/rollover.go

```diff
@@ -0,0 +1,144 @@
+package rollover
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/docker/docker/daemon/logger"
+	"github.com/docker/docker/pkg/jsonlog"
+)
+
+// RolloverLogger is Logger implementation for rollover docker logging:
+// JSON objects to file
+type RolloverLogger struct {
+	buf      *bytes.Buffer
+	f        *os.File // store for closing
+	capacity int64
+	n        int
+}
+
+func New(filename string, capacity string, n int) (logger.Logger, error) {
+	var capVal int64
+	if capacity != "" {
+		var multiplier int64
+		capString := capacity[:len(capacity)-1]
+		switch capacity[len(capacity)-1:] {
+		case "K":
+			multiplier = 1024
+		case "k":
+			multiplier = 1024
+		case "M":
+			multiplier = 1024 * 1024
+		case "m":
+			multiplier = 1024 * 1024
+		case "G":
+			multiplier = 1024 * 1024 * 1024
+		case "g":
+			multiplier = 1024 * 1024 * 1024
+		default:
+			multiplier = 1
+			capString = capacity
+		}
+		capVal, err := strconv.ParseInt(capString, 10, 64)
+		if err != nil {
+			return nil, err
+		}
+		capVal *= multiplier
+	} else {
+		capVal = -1
+	}
+	log, err := os.OpenFile(filename, os.O_WRONLY|os.O_APPEND|os.O_CREATE, 0600)
+	if err != nil {
+		return nil, err
+	}
+	return &RolloverLogger{
+		f:        log,
+		buf:      bytes.NewBuffer(nil),
+		capacity: capVal,
+		n:        n,
+	}, nil
+}
+
+// Log converts logger.Message to jsonlog.JSONLog and serializes it to file
+func (l *RolloverLogger) Log(msg *logger.Message) error {
+	if err := (&jsonlog.JSONLog{Log: string(msg.Line) + "\n", Stream: msg.Source, Created: msg.Timestamp}).MarshalJSONBuf(l.buf); err != nil {
+		return err
+	}
+	l.buf.WriteByte('\n')
+	_, err := writeLog(l)
+	return err
+}
+
+func writeLog(l *RolloverLogger) (int64, error) {
+	if l.capacity == -1 {
+		return l.buf.WriteTo(l.f)
+	}
+	meta, err := l.f.Stat()
```

Comment:
I'd be concerned about the number of stats you would have to do.  Not sure if this is going to be a problem.